#### Q1. MLflow version

In [3]:
import mlflow
print(mlflow.__version__)

1.26.0


#### Q.2 Solution Command:
python preprocess_data.py --raw_data_path /home/avijit/Desktop/Codes/mlops/Coursework/notebooks/data --dest_path ./output/

This saves 4 files in the output folder which are : 1.dv.pkl ,  2.train.pkl,   3.test.pkl,   4. valid.pkl

In [9]:
! python preprocess_data.py --raw_data_path /home/avijit/Desktop/Codes/mlops/Coursework/notebooks/data --dest_path ./output/

In [14]:
! ls ~/Desktop/Codes/mlops/Coursework/notebooks/homework/output

dv.pkl	test.pkl  train.pkl  valid.pkl


#### Q.3Train a model with autolog
How many parameters are automatically logged by MLflow?

In [19]:
! python train.py --data_path ~/Desktop/Codes/mlops/Coursework/notebooks/homework/output

2022/05/31 00:10:42 WARNING mlflow.utils.autologging_utils: You are using an unsupported version of sklearn. If you encounter errors during autologging, try upgrading / downgrading sklearn to a supported version, or try upgrading MLflow.
2022/05/31 00:11:28 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "/home/avijit/anaconda3/envs/exp-tracking-env/lib/python3.9/site-packages/_distutils_hack/__init__.py:30: UserWarning: Setuptools is replacing distutils."


Ans: 17 Parameters were stored. Which are

Name	            Value <br>
1.bootstrap	         True<br>
2.ccp_alpha	         0.0<br>
3.criterion	         squared_error<br>
4.max_depth	         10<br>
5.max_features	     1.0<br>
6.max_leaf_nodes	     None<br>
7.max_samples	None<br>
8.min_impurity_decrease	0.0<br>
9.min_samples_leaf	1<br>
10.min_samples_split	2<br>
11.min_weight_fraction_leaf	0.0<br>
12.n_estimators	100<br>
13.n_jobs	None<br>
14.oob_score	False<br>
15.random_state	0<br>
16.verbose	0<br>
17.warm_start	False<br>

#### Q4. Launching the tracking server locally

Ans: default-artifact-root

In [23]:
! mlflow ui --backend-store-uri sqlite:///mlflow_hw.db --default-artifact-root ~/Notebooks/week02/artifacts

[2022-05-31 02:12:26 +0530] [27631] [INFO] Starting gunicorn 20.1.0
[2022-05-31 02:12:26 +0530] [27631] [INFO] Listening at: http://127.0.0.1:5000 (27631)
[2022-05-31 02:12:26 +0530] [27631] [INFO] Using worker: sync
[2022-05-31 02:12:26 +0530] [27632] [INFO] Booting worker with pid: 27632
^C
[2022-05-31 02:12:51 +0530] [27631] [INFO] Handling signal: int
[2022-05-31 02:12:51 +0530] [27632] [INFO] Worker exiting (pid: 27632)


#### Q5. Tuning the hyperparameter of the model

In [24]:
! python hpo.py 

2022/05/31 02:20:53 INFO mlflow.tracking.fluent: Experiment with name 'random-forest-hyperopt' does not exist. Creating a new experiment.
100%|█████████| 50/50 [11:17<00:00, 13.55s/trial, best loss: 6.6284257482044735]


Ans: Best Validation RMSE: 6.628

#### Q6. Promoting best model to model registry

In [26]:
! python register_model.py

2022/05/31 02:52:19 WARNING mlflow.utils.autologging_utils: You are using an unsupported version of sklearn. If you encounter errors during autologging, try upgrading / downgrading sklearn to a supported version, or try upgrading MLflow.
2022/05/31 02:52:49 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "/home/avijit/anaconda3/envs/exp-tracking-env/lib/python3.9/site-packages/_distutils_hack/__init__.py:30: UserWarning: Setuptools is replacing distutils."
Model to be registered
Run id: 0108647402c84b518d07d5c6e049c14a, test_rmse: 6.5479
Successfully registered model 'EXPERIMENT_NAME'.
2022/05/31 02:54:42 INFO mlflow.tracking._model_registry.client: Waiting up to 300 seconds for model version to finish creation.                     Model name: EXPERIMENT_NAME, version 1
Created version '1' of model 'EXPERIMENT_NAME'.


Ans: Test RMSE On Best Model : 6.55